Refactoring LayoutContext and implementing with product editor

joelclimbsthings opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

As per the issue, we occasionally have the need to know "where we are" in the application, when using common components across different sections. Without complex prop drilling we can use the `LayoutContext` component. In this PR I've moved this to the `admin-layout` package, refactored the component, updated references, and implemented on the blocks editor.

Closes #37740 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout branch and enable the `new-product-management-experience` feature flag.
2. Run `localStorage.setItem( 'debug', 'wc-admin:tracks' );` in the console.
3. Navigate to WooCommerce -> Home
4. Ensure the `wcadmin_tasklist_view` event fires with the property `context: woocommerce-home`
5. Click on "Add Product" item in task list, and ensure that `wcadmin_tasklist_click` event fires with the same `context` property.
6. Navigate to the Customers screen and click the "Finish Setup" button in the Activity Panel to display the task list in the side bar.
7. Ensure the `wcadmin_tasklist_view` event fires, but this time the `context` property should be `woocommerce-analytics-customers/activity-panel`
8. Click the "Add a product" task again and ensure the `wcadmin_tasklist_click` has the same context property.
9. Navigate to Products -> Add New, and you should see the old React product editor.
10. Click the "More" menu within the Header and click "Share feedback"
11. Enter any information and click "Share"
12. The `wcadmin_ces_feedback` even should fire with the `block_editor: false` property.
13. Enable the `block-editor-feature-enabled` feature flag, and refresh the product editor to see the new product block editor.
14. Hit the more menu again, and complete the "Share feedback" modal again.
15. After clicking "Share" you should see the same `wcadmin_ces_feedback` event, but with `block_editor` equal to `true`.

<!-- End testing instructions -->

joelclimbsthings committed:
Refactoring layoutContext, moving to admin-layout package, and updating instances

joelclimbsthings committed:
Adding layout context for block editor and flagging on ces feedback event

joelclimbsthings committed:
Adding changelogs

joelclimbsthings committed:
Adding decendentOf shortcut function

joelclimbsthings committed:
Replacing error with warn due to test issues

joelclimbsthings committed:
Fixing tests and misspelled function name

joshuatf commented on packages/js/product-editor/src/components/editor/editor.tsx:
```diff
@@ -41,30 +45,42 @@ type EditorProps = {
-							}
-						/>
+		<LayoutContextProvider
+			value={ updateLayoutPath( 'product-block-editor' ) }
```
Since this would run on every render, should we call `updateLayoutPath` inside a `useEffect` or `useMemo`?

joelclimbsthings replied:
Yeahhh, most likely. Since this function is idempotent and non-complex I liked the elegance of being able to plug the helper function directly into the context, but it would be more efficient this way. I'll make the change.

joelclimbsthings replied:
I do wonder if this is necessary given some things I've read recently regarding [whether you really need an effect](https://react.dev/learn/you-might-not-need-an-effect#caching-expensive-calculations), which seems relevant to this kind of thing. Curious as to the thoughts of others on the team, but I'll commit this change for now.

joelclimbsthings replied:
Done in https://github.com/woocommerce/woocommerce/pull/37720/commits/6e5e0be29882fb42696e5e71a4fa2429483675fc

joshuatf commented on plugins/woocommerce-admin/client/activity-panel/activity-panel.js:
```diff
@@ -457,7 +450,7 @@ export const ActivityPanel = ( { isEmbedded, query } ) => {
 
 	return (
-		<LayoutContext.Provider value={ updatedLayoutContext }>
+		<LayoutContextProvider value={ updateLayoutPath( 'activity-panel' ) }>
```
Same question here.

joshuatf commented on packages/js/admin-layout/src/components/LayoutContext/layout-context.tsx:
```diff
@@ -0,0 +1,57 @@
+	layoutString: string;
+	updateLayoutPath: ( item: string ) => LayoutContextType;
+	layoutPath: string[];
+	descendantOf: ( item: string ) => boolean;
```
Nitpicky, but:
```suggestion
	isDescendantOf: ( item: string ) => boolean;
```

joelclimbsthings replied:
Works for me 👍🏻

joshuatf commented on packages/js/admin-layout/src/components/LayoutContext/layout-context.tsx:
```diff
@@ -0,0 +1,57 @@
+export type LayoutContextType = {
+	layoutString: string;
+	updateLayoutPath: ( item: string ) => LayoutContextType;
+	layoutPath: string[];
```
This and `layoutString` took me a minute to parse.  I wonder if something like `layoutParts`, `layoutBreadcrumbs` or `layoutHierarchy` might make this more clear.  (I was thinking `path` as in a string portion of the URL).

Similarly for `layoutString` this could just be `layout` if we make the above change.

joelclimbsthings replied:
Thanks @joshuatf , I made the following changes:
```
layoutPath -> layoutParts
updateLayoutPath -> extendLayout
descendantOf -> isDescendantOf
```

I left `layoutString` as it is for now, as I feel that `layout` doesn't quite work for it in my mind, and I'd like the name to be a bit more unique that just `layout`. Definitely open to other suggestions if you feel the need.

joshuatf reviewed (commented):
Nice work, Joel!  Left a few nitpicky naming comments, but overall this is looking pretty good to me.

joelclimbsthings committed:
Adding useMemo to make updateLayoutContext calls more efficient

joelclimbsthings committed:
Renaming descendantOf to isDescendantOf

joelclimbsthings committed:
Renaming layout context helpers

joelclimbsthings committed:
Adding useExtendLayout() helpoer

joelclimbsthings committed:
Fixing tests

joelclimbsthings commented:
Thanks @joshuatf , I made the requested changes as detailed in [this comment](https://github.com/woocommerce/woocommerce/pull/37720#discussion_r1175819647).

I also added a `useExtendLayout()` hook as a quick shortcut for memoizing and such when extending the layout, and implemented where appropriate. Let me know what you think!

joshuatf commented on plugins/woocommerce-admin/client/products/fills/more-menu-items/feedback-menu-item.tsx:
```diff
@@ -33,10 +35,15 @@ export const FeedbackMenuItem = ( { onClose }: { onClose: () => void } ) => {
 						icon: <span>🌟</span>,
+					},
+					{
+						block_editor: isDescendantOf( 'product-block-editor' ),
```
Any reason we don't continue to use the `context` property?  This seems like it might be an easier and more consistent way of checking context as we continue to iterate on the experience.

joelclimbsthings replied:
@joshuatf By the use of the `context` property, do you mean assigning that to the `layoutString` value, or do you just mean using a property called `context` here that provides a string that indicates the placement? 

For instance, manually setting `{ context: 'product-block-editor' }` versus just setting it to the current value of `layoutString` (as we've done elsewhere). 

The reason I didn't set it equal to `layoutString` is that it's a bit more fragile, as I believe you brought up on the original layoutContext PR a while back. If someone adds another provider the string will change and potentially break that implementation. 

I'm receptive to making this something along of the lines of `context: 'product-block-editor'` as mentioned instead of a boolean value, although then I actually have to come up with a name for the non-block react product editor for it's context value 😢 .

joshuatf replied:
>The reason I didn't set it equal to layoutString is that it's a bit more fragile, as I believe you brought up on the original layoutContext PR a while back. If someone adds another provider the string will change and potentially break that implementation.

Ya, that's a good point.  I think what you have here is good, no reason to introduce complications to it for minimal benefits.

joshuatf reviewed (approved):
Testing well and pre-approving!

Left one comment/question about the event property being used, but not overly opinionated on this and not sure if there is other context around this.
